shared/src/commonMain/kotlin/io/customer/shared/database/QueryHelper.kt

```diff
@@ -0,0 +1,264 @@
+    private val dateTimeUtil: DateTimeUtil,
+    private val jsonAdapter: JsonAdapter,
+    private val platformUtil: PlatformUtil,
+    private val workspace: Workspace,
+    private val backgroundQueueConfig: BackgroundQueueConfig,
+    private val trackingTaskQueries: TrackingTaskQueries,
+) : QueryHelper {
+    private val selectAllPendingQuery = trackingTaskQueries.selectAllPendingTasks(
+        status = listOf(QueueTaskStatus.PENDING, QueueTaskStatus.FAILED),
+        limit = backgroundQueueConfig.batchTasksMax.toLong(),
+        siteId = workspace.siteId,
+    )
+
+    /**
+     * Runs the given block in database transaction synchronously. The method returns result
+     * instantly and should only be called from background dispatcher.
+     */
+    private fun <Result : Any?> runInTransaction(
+        block: TransactionWithReturn<Result>.() -> Result,
+    ): Result {
+        return trackingTaskQueries.transactionWithResult(noEnclosing = false) { block() }
+    }
+
+    /**
+     * Runs the given block in database transaction asynchronously. The method does not return
+     * any result and should only be called where receiving result is not required.
+     */
+    private fun runInTransactionAsync(
+        block: TransactionWithReturn<Unit>.() -> Unit,
+    ) = runSuspended { runInTransaction(block = block) }
+
+    /**
+     * Looks for unsent duplicate tasks that can be merged and may not be required to be sent as
+     * multiple events e.g. Identify, Add Device, etc.
+     */
+    @WithinTransaction
+    private fun Task.getDuplicateOrNull(): TrackingTask? = kotlin.runCatching {
+        return@runCatching if (activity.canBeMerged()) {
+            trackingTaskQueries.selectByType(
+                type = activity.type,
+                siteId = workspace.siteId,
+            ).executeAsOneOrNull()?.takeUnless { task ->
+                task.queueTaskStatus in listOf(
+                    QueueTaskStatus.QUEUED,
+                    QueueTaskStatus.SENDING,
+                    QueueTaskStatus.SENT,
+                )
+            }
+        } else null
+    }.getOrNull()
+
+    /**
+     * Merges activities that are merge-able, e.g. attributes, etc.
+     */
+    @WithinTransaction
+    private fun TrackingTask.mergeActivities(activity: Activity): Activity? {
+        var mergedActivity: Activity? = null
+        kotlin.runCatching {
+            activity.merge(other = jsonAdapter.fromJSON(Activity::class, activityJson))
+        }.fold(
+            onSuccess = { value -> mergedActivity = value },
+            onFailure = { ex ->
+                logger.fatal("Failed to parse activity ${type}, model version ${activityModelVersion}. Reason: ${ex.message}")
+            },
+        )
+        return mergedActivity
+    }
+
+    /**
+     * Updates status of tasks with the given values.
+     *
+     * For private use only, should only be called within database transaction.
+     */
+    @WithinTransaction
```

Comment:
This annotation (`WithinTransaction`) too, was intended to be deleted later. And I would still probably vote for its deletion as I don't see it adding much value. But just asking if anyone has a different opinion.

Reply:
Add it once it has value. Otherwise, let's delete.

Reply:
Sure, deleting them too.